File: tools/run_state/migrate.py
"""One-shot migrators for legacy .prove/runs layout.

Converts the pre-JSON structure::

    .prove/runs/<branch>/<slug>/
      PRD.md
      TASK_PLAN.md
      PROGRESS.md           (optional)
      dispatch-state.json   (optional)
      reports/              (preserved as-is)

into the JSON-first structure::

    .prove/runs/<branch>/<slug>/
      prd.json
      plan.json
      state.json
      state.json.lock
      reports/              (new JSON reports added alongside legacy files)

Markdown parsing is deliberately tolerant — we extract what we can and
preserve the original body under ``body_markdown`` / ``description`` so
no information is lost. The legacy files are NOT deleted automatically;
the caller decides via ``--prune``.
"""

from __future__ import annotations

import dataclasses
import json
import re
from pathlib import Path
from typing import Iterable

from tools.run_state import CURRENT_SCHEMA_VERSION
from tools.run_state.state import (
    RunPaths,
    new_plan,
    new_prd,
    new_state,
    _write_json_atomic,
    utcnow_iso,
)


# Title line: "# Task Plan: ..." or "# <anything>"
_H1_RE = re.compile(r"^#\s+(.+)$", re.MULTILINE)

# "### Task 1.2: Something"
_TASK_RE = re.compile(
    r"^###\s+Task\s+(\d+(?:\.\d+)+):\s*(.+?)\s*$", re.MULTILINE
)

# "#### Step 1.2.3: Something" (optional sub-steps)
_STEP_RE = re.compile(
    r"^####\s+Step\s+(\d+(?:\.\d+)+):\s*(.+?)\s*$", re.MULTILINE
)

# "**Worktree:** /path/to/worktree"
_WORKTREE_RE = re.compile(r"^\*\*Worktree:\*\*\s*(.+?)\s*$", re.MULTILINE)
_BRANCH_RE = re.compile(r"^\*\*Branch:\*\*\s*(.+?)\s*$", re.MULTILINE)
_DEPS_RE = re.compile(
    r"^\*\*(?:Depends on|Dependencies):\*\*\s*(.+?)\s*$", re.MULTILINE
)


@dataclasses.dataclass
class MigrationResult:
    run_dir: Path
    prd_written: bool
    plan_written: bool
    state_written: bool
    tasks_found: int
    steps_found: int


# --------------------------------------------------------------------------
# PRD
# --------------------------------------------------------------------------


def parse_prd_md(text: str) -> dict:
    """Parse a legacy PRD.md into the prd.json shape (loose heuristics)."""
    title_match = _H1_RE.search(text)
    title = title_match.group(1).strip() if title_match else "Untitled Run"

    sections = _split_sections(text)
    context = _first_present(sections, ("Context", "Problem", "Background", "Summary"))
    goals = _extract_bullets(_first_present(sections, ("Goals", "Objectives")) or "")
    in_scope = _extract_bullets(_first_present(sections, ("In Scope", "Scope / In")) or "")
    out_scope = _extract_bullets(_first_present(sections, ("Out of Scope", "Out-of-Scope")) or "")
    acceptance = _extract_bullets(
        _first_present(sections, ("Acceptance Criteria", "Acceptance")) or ""
    )
    test_strategy = _first_present(sections, ("Test Strategy", "Testing", "Tests")) or ""

    return new_prd(
        title=title,
        context=context or "",
        goals=goals,
        scope={"in": in_scope, "out": out_scope},
        acceptance_criteria=acceptance,
        test_strategy=test_strategy,
        body_markdown=text.strip(),
    )


# --------------------------------------------------------------------------
# Plan
# --------------------------------------------------------------------------


def parse_plan_md(text: str) -> dict:
    """Parse a legacy TASK_PLAN.md into the plan.json shape."""
    # Split into task chunks by _TASK_RE matches; each task chunk spans from
    # its own match.start() to the next task's match.start() (or EOF).
    matches = list(_TASK_RE.finditer(text))
    tasks: list[dict] = []
    for i, m in enumerate(matches):
        start = m.end()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
        task_id = m.group(1).strip()
        title = m.group(2).strip()
        body = text[start:end].strip()

        wave = int(task_id.split(".")[0])

        wt_match = _WORKTREE_RE.search(body)
        br_match = _BRANCH_RE.search(body)
        deps_match = _DEPS_RE.search(body)
        deps = [
            d.strip()
            for d in (deps_match.group(1).split(",") if deps_match else [])
            if d.strip()
        ]

        # Sub-steps inside the task body (optional)
        step_matches = list(_STEP_RE.finditer(body))
        steps: list[dict] = []
        if step_matches:
            for j, sm in enumerate(step_matches):
                s_start = sm.end()
                s_end = (
                    step_matches[j + 1].start()
                    if j + 1 < len(step_matches)
                    else len(body)
                )
                s_id = sm.group(1).strip()
                s_title = sm.group(2).strip()
                s_desc = body[s_start:s_end].strip()
                steps.append(
                    {
                        "id": s_id,
                        "title": s_title,
                        "description": s_desc,
                        "acceptance_criteria": [],
                    }
                )
        else:
            # One implicit step — whole task body is the step description
            steps.append(
                {
                    "id": f"{task_id}.1",
                    "title": title,
                    "description": body,
                    "acceptance_criteria": [],
                }
            )

        tasks.append(
            {
                "id": task_id,
                "title": title,
                "wave": wave,
                "deps": deps,
                "description": body,
                "acceptance_criteria": [],
                "worktree": {
                    "path": wt_match.group(1).strip() if wt_match else "",
                    "branch": br_match.group(1).strip() if br_match else "",
                },
                "steps": steps,
            }
        )

    mode = "full" if any(int(t["wave"]) > 1 for t in tasks) else "simple"
    return new_plan(tasks=tasks, mode=mode)


# --------------------------------------------------------------------------
# State (from PROGRESS.md checklist + plan)
# --------------------------------------------------------------------------


_CHECK_TASK_RE = re.compile(
    r"^- \[(?P<mark>[ x!H~\-])\]\s+(?:Task\s+)?(?P<id>\d+(?:\.\d+)+)", re.MULTILINE
)


def derive_state_from_progress(progress_text: str, plan: dict, slug: str, branch: str) -> dict:
    """Best-effort translation of a PROGRESS.md checklist into state.json.

    Unmatched tasks keep their default ``pending`` status.
    """
    state = new_state(slug=slug, branch=branch, plan=plan)

    # Parse checkmarks into a {task_id: status} map
    statuses: dict[str, str] = {}
    for m in _CHECK_TASK_RE.finditer(progress_text):
        mark = m.group("mark")
        tid = m.group("id")
        statuses[tid] = _mark_to_status(mark)

    any_in_progress = False
    for task in state.get("tasks", []):
        s = statuses.get(task["id"])
        if not s:
            continue
        task["status"] = s
        # Propagate a sensible default to steps: completed tasks mark steps completed,
        # in_progress leaves steps pending, failed marks first step failed, halted likewise.
        if s == "completed":
            for step in task["steps"]:
                step["status"] = "completed"
                step["ended_at"] = utcnow_iso()
            task["ended_at"] = utcnow_iso()
        elif s == "in_progress":
            any_in_progress = True
            # leave steps pending; caller's hot path will update as it runs
        elif s in ("failed", "halted"):
            task["ended_at"] = utcnow_iso()

    if any_in_progress:
        state["run_status"] = "running"
    elif all(t["status"] == "completed" for t in state["tasks"]) and state["tasks"]:
        state["run_status"] = "completed"
        state["ended_at"] = utcnow_iso()

    return state


def _mark_to_status(mark: str) -> str:
    return {
        " ": "pending",
        "x": "completed",
        "!": "failed",
        "H": "halted",
        "~": "in_progress",
        "-": "skipped",
    }.get(mark, "pending")


# --------------------------------------------------------------------------
# Migration driver
# --------------------------------------------------------------------------


def migrate_run(
    run_dir: Path,
    *,
    branch: str,
    slug: str,
    dry_run: bool = False,
    overwrite: bool = False,
) -> MigrationResult:
    """Convert a single run directory to the JSON-first layout."""
    prd_md = run_dir / "PRD.md"
    plan_md = run_dir / "TASK_PLAN.md"
    progress_md = run_dir / "PROGRESS.md"
    dispatch_legacy = run_dir / "dispatch-state.json"

    paths = RunPaths(
        root=run_dir,
        prd=run_dir / "prd.json",
        plan=run_dir / "plan.json",
        state=run_dir / "state.json",
        state_lock=run_dir / "state.json.lock",
        reports_dir=run_dir / "reports",
    )

    # PRD
    prd_written = False
    if prd_md.exists() and (not paths.prd.exists() or overwrite):
        prd = parse_prd_md(prd_md.read_text(encoding="utf-8"))
        if not dry_run:
            _write_json_atomic(paths.prd, prd)
        prd_written = True

    # Plan
    plan: dict | None = None
    plan_written = False
    if plan_md.exists() and (not paths.plan.exists() or overwrite):
        plan = parse_plan_md(plan_md.read_text(encoding="utf-8"))
        if not dry_run:
            _write_json_atomic(paths.plan, plan)
        plan_written = True
    elif paths.plan.exists():
        plan = json.loads(paths.plan.read_text(encoding="utf-8"))

    # State
    state_written = False
    if plan is not None and (not paths.state.exists() or overwrite):
        if progress_md.exists():
            state = derive_state_from_progress(
                progress_md.read_text(encoding="utf-8"),
                plan,
                slug=slug,
                branch=branch,
            )
        else:
            state = new_state(slug=slug, branch=branch, plan=plan)

        # Fold legacy dispatch-state.json into state.dispatch
        if dispatch_legacy.exists():
            try:
                legacy = json.loads(dispatch_legacy.read_text(encoding="utf-8"))
                if isinstance(legacy.get("dispatched"), list):
                    state.setdefault("dispatch", {"dispatched": []})
                    state["dispatch"]["dispatched"].extend(legacy["dispatched"])
            except json.JSONDecodeError:
                pass

        if not dry_run:
            _write_json_atomic(paths.state, state)
        state_written = True

    tasks_found = len(plan["tasks"]) if plan else 0
    steps_found = sum(len(t.get("steps", [])) for t in (plan["tasks"] if plan else []))

    return MigrationResult(
        run_dir=run_dir,
        prd_written=prd_written,
        plan_written=plan_written,
        state_written=state_written,
        tasks_found=tasks_found,
        steps_found=steps_found,
    )


def migrate_all(runs_root: Path, *, dry_run: bool = False, overwrite: bool = False) -> list[MigrationResult]:
    """Walk ``.prove/runs/`` and migrate every leaf run directory found.

    A leaf is any directory containing ``TASK_PLAN.md`` or ``PRD.md``. The
    branch is the first path component below ``runs_root``; the slug is
    the run directory name.
    """
    results: list[MigrationResult] = []
    if not runs_root.exists():
        return results

    for path in sorted(_iter_run_dirs(runs_root)):
        rel = path.relative_to(runs_root).parts
        if len(rel) == 1:
            # legacy top-level run (no branch namespace)
            branch = "main"
            slug = rel[0]
        else:
            branch = rel[0]
            slug = rel[-1]
        results.append(
            migrate_run(path, branch=branch, slug=slug, dry_run=dry_run, overwrite=overwrite)
        )
    return results


def _iter_run_dirs(runs_root: Path) -> Iterable[Path]:
    for entry in runs_root.rglob("*"):
        if entry.is_dir() and (
            (entry / "TASK_PLAN.md").exists() or (entry / "PRD.md").exists()
        ):
            yield entry


# --------------------------------------------------------------------------
# Markdown section helpers
# --------------------------------------------------------------------------


_SECTION_RE = re.compile(r"^##\s+(.+?)\s*$", re.MULTILINE)


def _split_sections(text: str) -> dict[str, str]:
    """Split markdown by ## headings; return {heading: body}."""
    out: dict[str, str] = {}
    matches = list(_SECTION_RE.finditer(text))
    for i, m in enumerate(matches):
        heading = m.group(1).strip()
        start = m.end()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
        out[heading] = text[start:end].strip()
    return out


def _first_present(sections: dict[str, str], candidates: Iterable[str]) -> str | None:
    for name in candidates:
        if name in sections:
            return sections[name]
    # Case-insensitive fallback
    lower = {k.lower(): v for k, v in sections.items()}
    for name in candidates:
        if name.lower() in lower:
            return lower[name.lower()]
    return None


_BULLET_RE = re.compile(r"^\s*[-*]\s+(.+?)\s*$", re.MULTILINE)


def _extract_bullets(text: str) -> list[str]:
    return [m.group(1).strip() for m in _BULLET_RE.finditer(text)]

File: packages/cli/src/topics/run-state/migrate.ts
/**
 * One-shot migrators for legacy `.prove/runs/<branch>/<slug>/` layouts.
 *
 * Ported 1:1 from `tools/run_state/migrate.py`. Converts the legacy
 * markdown-first structure
 *
 *   .prove/runs/<branch>/<slug>/
 *     PRD.md
 *     TASK_PLAN.md
 *     PROGRESS.md           (optional)
 *     dispatch-state.json   (optional)
 *     reports/              (preserved as-is)
 *
 * into the JSON-first shape
 *
 *   .prove/runs/<branch>/<slug>/
 *     prd.json
 *     plan.json
 *     state.json
 *     reports/              (new JSON reports added alongside legacy files)
 *
 * Markdown parsing is deliberately tolerant — extract what we can and
 * preserve the original body under `body_markdown` / `description` so no
 * information is lost. Legacy files are NOT deleted; callers decide.
 *
 * This module is standalone on purpose — it inlines `newPrd` / `newPlan`
 * / `newState` / `utcnowIso` / `_writeJsonAtomic` from `state.py` so the
 * migrate entrypoint has zero dependency on the still-in-flight state
 * port (Task 2). The factory helpers mirror Python construction order
 * exactly so JSON key order matches byte-for-byte.
 *
 * Note: the task brief described a schema-version migration chain. The
 * actual Python source is a markdown→JSON converter, not a chain — there
 * is one schema version (v1) and no `MIGRATIONS` registry. This port
 * preserves Python semantics faithfully; see `tools/run_state/migrate.py`
 * and `.prove/decisions/2026-04-17-prove-runs-json-first.md` for intent.
 */

import { existsSync, readFileSync, readdirSync, statSync } from 'node:fs';
import { join, relative, resolve, sep } from 'node:path';
import type { RunPathsData } from './paths';
import {
  type DispatchLedger,
  type PlanData,
  type PrdData,
  type StateData,
  type TaskData,
  _clock,
  deepCloneJson,
  defaultsFromSchema,
  newPlan as stateNewPlan,
  newPrd as stateNewPrd,
  newState as stateNewState,
  utcnowIso,
  writeJsonAtomic,
} from './state';

// --------------------------------------------------------------------------
// Re-exports — a single source of truth lives in state.ts. The alias
// wrappers below keep the `migrate.*` import surface stable for callers
// (Python tests used a local `migrate.utcnow_iso` monkeypatch, the TS
// equivalent just reads through to `state._clock`).
// --------------------------------------------------------------------------

export { _clock, deepCloneJson, defaultsFromSchema, utcnowIso, writeJsonAtomic };

/**
 * Raised when the legacy markdown input cannot be translated into a valid
 * JSON artifact — typically because the parsed plan is missing required
 * fields (`tasks` array, task `id`, step `id`). Surfaces the offending
 * field so callers can point at the bad source file.
 */
export class MigrationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'MigrationError';
  }
}

/**
 * Build a fresh prd.json payload. Thin passthrough to the canonical factory
 * in `state.ts`; exported from this module so tests and migrate call sites
 * share one construction path (and thus one key order).
 */
export function newPrd(title: string, overrides: Record<string, unknown> = {}): PrdData {
  return stateNewPrd(title, overrides);
}

export function newPlan(tasks: PlanData['tasks'], mode = 'simple'): PlanData {
  return stateNewPlan(tasks, mode);
}

/**
 * Markdown-tolerant `newState` wrapper. Accepts the untyped dict shape
 * the markdown parsers produce; validates the minimal plan shape then
 * delegates to the canonical builder in `state.ts` so key order / on-disk
 * bytes stay identical.
 */
export function newState(slug: string, branch: string, plan: Record<string, unknown>): StateData {
  return stateNewState(slug, branch, coercePlan(plan));
}

/**
 * Narrow a markdown-parsed dict into a `PlanData` shape. The parsers in
 * this file build plans field-by-field, so `tasks` is always present when
 * the markdown is well-formed; this guard raises `MigrationError` rather
 * than allowing a silently-malformed plan to flow into `state.json`.
 */
function coercePlan(plan: Record<string, unknown>): PlanData {
  if (!Array.isArray(plan.tasks)) {
    throw new MigrationError(
      "plan is missing 'tasks' array (expected PlanData shape from parsePlanMd)",
    );
  }
  return plan as unknown as PlanData;
}

// --------------------------------------------------------------------------
// Markdown regex primitives — mirrors migrate.py module-level regexes.
//
// WARNING: these regexes all carry the `g` flag and are safe only because
// callers use `String.prototype.matchAll` (stateless). Do NOT switch to
// `.exec()` in a loop — the global flag causes `lastIndex` to carry
// between calls and produces silent misparses on the second pass.
// --------------------------------------------------------------------------

// H1 title: "# Task Plan: ..." or "# <anything>"
const H1_RE = /^#\s+(.+)$/m;

// "### Task 1.2: Something"
const TASK_RE = /^###\s+Task\s+(\d+(?:\.\d+)+):\s*(.+?)\s*$/gm;

// "#### Step 1.2.3: Something" (optional sub-steps)
const STEP_RE = /^####\s+Step\s+(\d+(?:\.\d+)+):\s*(.+?)\s*$/gm;

// "**Worktree:** /path/to/worktree"
const WORKTREE_RE = /^\*\*Worktree:\*\*\s*(.+?)\s*$/m;
const BRANCH_RE = /^\*\*Branch:\*\*\s*(.+?)\s*$/m;
const DEPS_RE = /^\*\*(?:Depends on|Dependencies):\*\*\s*(.+?)\s*$/m;

// "## Section Heading"
const SECTION_RE = /^##\s+(.+?)\s*$/gm;

// "- bullet" or "* bullet"
const BULLET_RE = /^\s*[-*]\s+(.+?)\s*$/gm;

// Progress line: "- [x] Task 1.2 ..."
const CHECK_TASK_RE = /^- \[([ x!H~\-])\]\s+(?:Task\s+)?(\d+(?:\.\d+)+)/gm;

// --------------------------------------------------------------------------
// PRD
// --------------------------------------------------------------------------

/** Parse a legacy PRD.md into the prd.json shape (loose heuristics). */
export function parsePrdMd(text: string): Record<string, unknown> {
  const titleMatch = text.match(H1_RE);
  const title = titleMatch?.[1] ? titleMatch[1].trim() : 'Untitled Run';

  const sections = splitSections(text);
  const context = firstPresent(sections, ['Context', 'Problem', 'Background', 'Summary']);
  const goals = extractBullets(firstPresent(sections, ['Goals', 'Objectives']) ?? '');
  const inScope = extractBullets(firstPresent(sections, ['In Scope', 'Scope / In']) ?? '');
  const outScope = extractBullets(firstPresent(sections, ['Out of Scope', 'Out-of-Scope']) ?? '');
  const acceptance = extractBullets(
    firstPresent(sections, ['Acceptance Criteria', 'Acceptance']) ?? '',
  );
  const testStrategy = firstPresent(sections, ['Test Strategy', 'Testing', 'Tests']) ?? '';

  return newPrd(title, {
    context: context ?? '',
    goals,
    scope: { in: inScope, out: outScope },
    acceptance_criteria: acceptance,
    test_strategy: testStrategy,
    body_markdown: text.trim(),
  });
}

// --------------------------------------------------------------------------
// Plan
// --------------------------------------------------------------------------

/** Parse a legacy TASK_PLAN.md into the plan.json shape. */
export function parsePlanMd(text: string): Record<string, unknown> {
  const matches = Array.from(text.matchAll(TASK_RE));
  const tasks: Record<string, unknown>[] = [];

  for (let i = 0; i < matches.length; i++) {
    const m = matches[i];
    if (!m || m.index === undefined) continue;
    const taskId = (m[1] ?? '').trim();
    const title = (m[2] ?? '').trim();
    const start = m.index + m[0].length;
    const nextMatch = matches[i + 1];
    const end = nextMatch && nextMatch.index !== undefined ? nextMatch.index : text.length;
    const body = text.slice(start, end).trim();

    const wave = Number.parseInt(taskId.split('.')[0] ?? '0', 10);

    const wtMatch = body.match(WORKTREE_RE);
    const brMatch = body.match(BRANCH_RE);
    const depsMatch = body.match(DEPS_RE);
    const deps = depsMatch?.[1]
      ? depsMatch[1]
          .split(',')
          .map((d) => d.trim())
          .filter((d) => d.length > 0)
      : [];

    // Sub-steps inside the task body (optional)
    const stepMatches = Array.from(body.matchAll(STEP_RE));
    const steps: Record<string, unknown>[] = [];

    if (stepMatches.length > 0) {
      for (let j = 0; j < stepMatches.length; j++) {
        const sm = stepMatches[j];
        if (!sm || sm.index === undefined) continue;
        const sId = (sm[1] ?? '').trim();
        const sTitle = (sm[2] ?? '').trim();
        const sStart = sm.index + sm[0].length;
        const nextStep = stepMatches[j + 1];
        const sEnd = nextStep && nextStep.index !== undefined ? nextStep.index : body.length;
        const sDesc = body.slice(sStart, sEnd).trim();
        steps.push({
          id: sId,
          title: sTitle,
          description: sDesc,
          acceptance_criteria: [],
        });
      }
    } else {
      // One implicit step — whole task body is the step description.
      steps.push({
        id: `${taskId}.1`,
        title,
        description: body,
        acceptance_criteria: [],
      });
    }

    tasks.push({
      id: taskId,
      title,
      wave,
      deps,
      description: body,
      acceptance_criteria: [],
      worktree: {
        path: wtMatch?.[1] ? wtMatch[1].trim() : '',
        branch: brMatch?.[1] ? brMatch[1].trim() : '',
      },
      steps,
    });
  }

  const mode = tasks.some((t) => Number(t.wave) > 1) ? 'full' : 'simple';
  // Markdown parsing built each task field-by-field against PlanTaskInput;
  // structural match is verified by the parity fixtures.
  return newPlan(tasks as unknown as PlanData['tasks'], mode);
}

// --------------------------------------------------------------------------
// State (from PROGRESS.md checklist + plan)
// --------------------------------------------------------------------------

/**
 * Best-effort translation of a PROGRESS.md checklist into state.json.
 * Unmatched tasks keep their default `pending` status.
 *
 * Input-shape discipline: `newState` guarantees a well-formed `StateData`
 * output so bracket-notation writes are replaced by typed field access —
 * the compiler now rejects value drift. Callers passing a malformed plan
 * surface as `MigrationError` from `newState`, never silent corruption.
 */
export function deriveStateFromProgress(
  progressText: string,
  plan: Record<string, unknown>,
  slug: string,
  branch: string,
): StateData {
  const state = newState(slug, branch, plan);

  // Parse checkmarks into a {taskId: status} map.
  const statuses: Record<string, ProgressMark> = {};
  for (const m of progressText.matchAll(CHECK_TASK_RE)) {
    const mark = m[1] ?? ' ';
    const tid = m[2];
    if (!tid) continue;
    statuses[tid] = markToStatus(mark);
  }

  let anyInProgress = false;

  for (const task of state.tasks) {
    const s = statuses[task.id];
    if (!s) continue;
    // 'skipped' is a step-only status and 'pending' is a no-op default —
    // filter to the four marks that legally transition a task.
    if (!TASK_MARKS.has(s)) continue;
    task.status = s as TaskData['status'];

    // Propagate a sensible default to steps: completed tasks mark steps
    // completed, in_progress leaves steps pending, failed/halted mark
    // the task end-time but leave steps for the caller.
    if (s === 'completed') {
      for (const step of task.steps) {
        step.status = 'completed';
        step.ended_at = utcnowIso();
      }
      task.ended_at = utcnowIso();
    } else if (s === 'in_progress') {
      anyInProgress = true;
    } else if (s === 'failed' || s === 'halted') {
      task.ended_at = utcnowIso();
    }
  }

  if (anyInProgress) {
    state.run_status = 'running';
  } else if (state.tasks.length > 0 && state.tasks.every((t) => t.status === 'completed')) {
    state.run_status = 'completed';
    state.ended_at = utcnowIso();
  }

  return state;
}

type ProgressMark = 'pending' | 'completed' | 'failed' | 'halted' | 'in_progress' | 'skipped';

/** Marks that legally land on `TaskData.status`. Other marks are ignored. */
const TASK_MARKS: ReadonlySet<ProgressMark> = new Set([
  'completed',
  'failed',
  'halted',
  'in_progress',
]);

function markToStatus(mark: string): ProgressMark {
  switch (mark) {
    case ' ':
      return 'pending';
    case 'x':
      return 'completed';
    case '!':
      return 'failed';
    case 'H':
      return 'halted';
    case '~':
      return 'in_progress';
    case '-':
      return 'skipped';
    default:
      return 'pending';
  }
}

// --------------------------------------------------------------------------
// Migration driver
// --------------------------------------------------------------------------

export interface MigrationResult {
  runDir: string;
  prdWritten: boolean;
  planWritten: boolean;
  stateWritten: boolean;
  tasksFound: number;
  stepsFound: number;
}

export interface MigrateRunOptions {
  branch: string;
  slug: string;
  dryRun?: boolean;
  overwrite?: boolean;
}

/** Convert a single run directory to the JSON-first layout. */
export function migrateRun(runDir: string, opts: MigrateRunOptions): MigrationResult {
  const { branch, slug, dryRun = false, overwrite = false } = opts;

  const prdMd = join(runDir, 'PRD.md');
  const planMd = join(runDir, 'TASK_PLAN.md');
  const progressMd = join(runDir, 'PROGRESS.md');
  const dispatchLegacy = join(runDir, 'dispatch-state.json');

  // `migrateRun` takes a pre-resolved run dir rather than `(runsRoot,
  // branch, slug)`, so we compute paths inline against the canonical
  // `RunPathsData` shape from `./paths` instead of calling `forRun`.
  const paths: RunPathsData = {
    root: runDir,
    prd: join(runDir, 'prd.json'),
    plan: join(runDir, 'plan.json'),
    state: join(runDir, 'state.json'),
    state_lock: join(runDir, 'state.json.lock'),
    reports_dir: join(runDir, 'reports'),
  };

  // PRD
  let prdWritten = false;
  if (existsSync(prdMd) && (!existsSync(paths.prd) || overwrite)) {
    const prd = parsePrdMd(readFileSync(prdMd, 'utf8'));
    if (!dryRun) writeJsonAtomic(paths.prd, prd);
    prdWritten = true;
  }

  // Plan
  let plan: Record<string, unknown> | null = null;
  let planWritten = false;
  if (existsSync(planMd) && (!existsSync(paths.plan) || overwrite)) {
    plan = parsePlanMd(readFileSync(planMd, 'utf8'));
    if (!dryRun) writeJsonAtomic(paths.plan, plan);
    planWritten = true;
  } else if (existsSync(paths.plan)) {
    plan = JSON.parse(readFileSync(paths.plan, 'utf8')) as Record<string, unknown>;
  }

  // State
  let stateWritten = false;
  if (plan !== null && (!existsSync(paths.state) || overwrite)) {
    const state = existsSync(progressMd)
      ? deriveStateFromProgress(readFileSync(progressMd, 'utf8'), plan, slug, branch)
      : newState(slug, branch, plan);

    // Fold legacy dispatch-state.json into state.dispatch.
    if (existsSync(dispatchLegacy)) {
      try {
        const legacy = JSON.parse(readFileSync(dispatchLegacy, 'utf8')) as Record<string, unknown>;
        if (Array.isArray(legacy.dispatched)) {
          // `newState` always seeds `state.dispatch = { dispatched: [] }`,
          // so no re-initialization guard is needed here.
          state.dispatch.dispatched = [
            ...state.dispatch.dispatched,
            ...(legacy.dispatched as DispatchLedger['dispatched']),
          ];
        }
      } catch {
        // Malformed legacy file — skip silently, mirrors Python's except JSONDecodeError.
      }
    }

    if (!dryRun) writeJsonAtomic(paths.state, state);
    stateWritten = true;
  }

  const planTasks = plan ? ((plan.tasks as unknown[]) ?? []) : [];
  const tasksFound = planTasks.length;
  let stepsFound = 0;
  for (const t of planTasks) {
    const steps = (t as Record<string, unknown>).steps;
    if (Array.isArray(steps)) stepsFound += steps.length;
  }

  return { runDir, prdWritten, planWritten, stateWritten, tasksFound, stepsFound };
}

export interface MigrateAllOptions {
  dryRun?: boolean;
  overwrite?: boolean;
}

/**
 * Walk `.prove/runs/` and migrate every leaf run directory found.
 * A leaf is any directory containing `TASK_PLAN.md` or `PRD.md`. The
 * branch is the first path component below `runsRoot`; the slug is the
 * run directory name.
 */
export function migrateAll(runsRoot: string, opts: MigrateAllOptions = {}): MigrationResult[] {
  const results: MigrationResult[] = [];
  if (!existsSync(runsRoot)) return results;

  const absRoot = resolve(runsRoot);
  const leaves = iterRunDirs(absRoot).sort();

  for (const path of leaves) {
    const rel = relative(absRoot, path).split(sep).filter(Boolean);
    let branch: string;
    let slug: string;
    if (rel.length === 1) {
      // legacy top-level run (no branch namespace)
      branch = 'main';
      slug = rel[0] ?? '';
    } else {
      branch = rel[0] ?? 'main';
      slug = rel[rel.length - 1] ?? '';
    }
    results.push(
      migrateRun(path, {
        branch,
        slug,
        dryRun: opts.dryRun,
        overwrite: opts.overwrite,
      }),
    );
  }

  return results;
}

function iterRunDirs(root: string): string[] {
  const out: string[] = [];
  walk(root, out);
  return out;
}

function walk(dir: string, acc: string[]): void {
  let entries: string[];
  try {
    entries = readdirSync(dir);
  } catch {
    return;
  }
  const hasLegacy = existsSync(join(dir, 'TASK_PLAN.md')) || existsSync(join(dir, 'PRD.md'));
  if (hasLegacy) acc.push(dir);
  for (const name of entries) {
    const child = join(dir, name);
    let st: ReturnType<typeof statSync>;
    try {
      st = statSync(child);
    } catch {
      continue;
    }
    if (st.isDirectory()) walk(child, acc);
  }
}

// --------------------------------------------------------------------------
// Markdown section helpers
// --------------------------------------------------------------------------

/** Split markdown by `##` headings; return `{heading: body}`. */
function splitSections(text: string): Record<string, string> {
  const out: Record<string, string> = {};
  const matches = Array.from(text.matchAll(SECTION_RE));
  for (let i = 0; i < matches.length; i++) {
    const m = matches[i];
    if (!m || m.index === undefined) continue;
    const heading = (m[1] ?? '').trim();
    const start = m.index + m[0].length;
    const next = matches[i + 1];
    const end = next && next.index !== undefined ? next.index : text.length;
    out[heading] = text.slice(start, end).trim();
  }
  return out;
}

function firstPresent(sections: Record<string, string>, candidates: string[]): string | null {
  for (const name of candidates) {
    const hit = sections[name];
    if (hit !== undefined) return hit;
  }
  // Case-insensitive fallback
  const lower: Record<string, string> = {};
  for (const [k, v] of Object.entries(sections)) {
    lower[k.toLowerCase()] = v;
  }
  for (const name of candidates) {
    const hit = lower[name.toLowerCase()];
    if (hit !== undefined) return hit;
  }
  return null;
}

function extractBullets(text: string): string[] {
  return Array.from(text.matchAll(BULLET_RE)).map((m) => (m[1] ?? '').trim());
}
